otherlibs/dune-rpc/private/procedures.ml

```diff
@@ -0,0 +1,366 @@
+open Types
+open Exported_types
+
+let id x = x
+
+module Public = struct
+  module Ping = struct
+    module V1 = struct
+      let version = 1
+
+      type req = unit
+
+      type resp = unit
+
+      type wire_req = unit
+
+      type wire_resp = unit
+
+      let req = Conv.unit
+
+      let resp = Conv.unit
+
+      let upgrade_req = id
+
+      let upgrade_resp = id
+
+      let downgrade_req = id
+
+      let downgrade_resp = id
+    end
+
+    let decl =
+      Decl.Request.make ~method_:"ping"
+        ~generations:[ Decl.Request.make_gen (module V1) ]
+  end
+
+  module Diagnostics = struct
+    module V1 = struct
+      let version = 1
+
+      type req = unit
+
+      type resp = Diagnostic.t list
+
+      type wire_req = unit
+
+      type wire_resp = Diagnostic.t list
+
+      let req = Conv.unit
+
+      let resp = Conv.list Diagnostic.sexp
+
+      let upgrade_req = id
+
+      let upgrade_resp = id
+
+      let downgrade_req = id
+
+      let downgrade_resp = id
+    end
+
+    let decl =
+      Decl.Request.make ~method_:"diagnostics"
+        ~generations:[ Decl.Request.make_gen (module V1) ]
+  end
+
+  module Subscribe_ = struct
+    module V1 = struct
```

Comment:
What about cleaning up all this `module V1` boilerplate with a functor?

Reply:
Or avoid first class modules altogether.

Reply:
I'm not sure what I like best here.

We can paper over a lot of the boilerplate by hiding it behind convenience functions (especially in cases where all the conversions are the identity), but ultimately it's all necessary (we always need all the conversion functions).

I don't feel strongly about packing the wire types with a GADT+record vs first class modules. Async_rpc uses a `Register` functor that stores the types in some global mutable table (which I also did initially), but this makes testing and reasoning about which functions are actually available quite difficult. I'll rewrite to get rid of the modules and see if it's actually better.